compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PromoteUsedTemporaries.ts

```diff
@@ -148,6 +152,243 @@ class CollectPromotableTemporaries extends ReactiveFunctionVisitor<State> {
+           * during codegen, then it can interpose between the defs and the uses of other temporaries.
+           * Since this instruction could potentially mutate those defs, it's not safe to relocate
+           * the definition of those temporaries to after this instruction. Mark all those temporaries
+           * as needing promotion, but don't promote them until we actually see them being used.
+           */
+          for (const [key, [ident, _]] of preEntries) {
+            state.set(key, [ident, true]);
+          }
+        }
+        break;
+      }
+      case 'DeclareContext':
+      case 'DeclareLocal': {
+        if (
+          instruction.value.lvalue.kind === 'Const' ||
+          instruction.value.lvalue.kind === 'HoistedConst'
+        ) {
+          this.#consts.add(instruction.value.lvalue.place.identifier.id);
+        }
+        super.visitInstruction(instruction, state);
+        break;
+      }
+      case 'LoadContext':
+      case 'LoadLocal': {
+        if (instruction.lvalue && instruction.lvalue.identifier.name === null) {
+          if (this.#consts.has(instruction.value.place.identifier.id)) {
+            this.#consts.add(instruction.lvalue.identifier.id);
+          }
+          state.set(instruction.lvalue.identifier.id, [
+            instruction.lvalue.identifier,
+            false,
+          ]);
+        }
+        super.visitInstruction(instruction, state);
+        break;
+      }
+      case 'PropertyLoad':
+      case 'ComputedLoad': {
+        if (instruction.lvalue && instruction.lvalue.identifier.name === null) {
+          state.set(instruction.lvalue.identifier.id, [
+            instruction.lvalue.identifier,
+            false,
+          ]);
+        }
+        super.visitInstruction(instruction, state);
+        break;
+      }
+      default: {
+        super.visitInstruction(instruction, state);
+      }
+    }
+  }
+
+  override traverseTerminal(
+    stmt: ReactiveTerminalStatement,
+    state: InterState,
+  ): void {
+    function copyState(): InterState {
+      const map: InterState = new Map();
+      state.forEach(([id, promo], key) => map.set(key, [id, promo]));
+      return map;
+    }
+    function mergeFrom(state2: InterState): void {
+      for (const [key, [ident, promo1]] of state2.entries()) {
+        const promo2 = state.get(key)?.[1] ?? false;
+        state.set(key, [ident, promo1 || promo2]);
+      }
+    }
+
+    const {terminal} = stmt;
+    if (terminal.id !== null) {
+      this.visitID(terminal.id, state);
+    }
+    switch (terminal.kind) {
+      case 'if': {
+        /*
+         * Visit if branches separately -- we don't want
+         * to promote a use in the alternate branch because
+         * of an interposing instruction in the consequent branch,
+         * because it's not actually interposing at runtime.
+         * After finishing the branches, merge their states.
+         */
```

Comment:
this makes sense, you want to merge the state for the outer block context, but don't want to merge the state of the locals per branch.

can you add some more fixtures to stress test this?

Reply:
It turns out I don't think I can construct an example that exercises this behavior--this would require a temporary to be defined before an if terminal and then used within one branch of the terminal.

I'm inclined to drop this code path because of this--even if I'm wrong and we can construct an example that exercises this, the only consequence would be an unnecessarily promoted temporary.

Reply:
Makes sense!